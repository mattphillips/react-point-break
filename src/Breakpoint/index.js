import { Component } from 'react';
import { func, object } from 'prop-types';

class Breakpoint extends Component {
  static propTypes = {
    children: func.isRequired,
    queries: object.isRequired // eslint-disable-line
  };

  static defaultProps = {
    queries: []
  };

  constructor(props) {
    super(props);
    this.state = Object.entries(props.queries).reduce((acc, [name]) => ({ ...acc, [name]: false }), {});
  }

  componentDidMount() {
    this.queries = Object.entries(this.props.queries).map(([name, query]) => {
      const mql = window.matchMedia(query);
      mql.addListener(this.update);
      return { mql, name };
    });

    this.update();
  }

  componentWillUnmount() {
    this.queries.forEach(({ mql }) => {
      mql.removeListener(this.update);
    });
  }

  update = () =>
    this.setState(this.queries.reduce((acc, { mql, name }) => ({ ...acc, [name]: mql.matches }), this.state));

  render() {
    return this.props.children(this.state);
  }
}

export default Breakpoint;
