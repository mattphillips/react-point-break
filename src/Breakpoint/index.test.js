import React from 'react';
import { mount } from 'enzyme';

import Breakpoint from './';

const noop = () => {};

const queries = {
  md: '(min-width: 801px) and (max-width: 900px)',
  lg: '(min-width: 901px)'
};

describe('<Breakpoint />', () => {
  it('renders with given queries set to false', () => {
    window.matchMedia = () => ({ addListener: noop });
    const component = mount(
      <Breakpoint queries={queries}>
        {({ md, lg }) => {
          if (lg) return <h1>Large</h1>;
          if (md) return <h1>Medium</h1>;
          return <h1>Small</h1>;
        }}
      </Breakpoint>
    );
    expect(component).toMatchSnapshot();
  });

  it('calls matchMedia with given queries and registers listeners when mounted and window is defined', () => {
    const addListener = jest.fn();
    const matchMedia = jest.fn(() => ({ addListener }));
    window.matchMedia = matchMedia;
    mount(
      <Breakpoint queries={queries}>
        {({ md, lg }) => {
          if (lg) return <h1>Large</h1>;
          if (md) return <h1>Medium</h1>;
          return <h1>Small</h1>;
        }}
      </Breakpoint>
    );
    expect(matchMedia).toHaveBeenCalledTimes(2);
    expect(matchMedia.mock.calls).toIncludeAllMembers([[queries.md], [queries.lg]]);
    expect(addListener).toHaveBeenCalledTimes(2);
  });

  it('calls remove listener when unmounted', () => {
    const addListener = jest.fn();
    const removeListener = jest.fn();
    const matchMedia = jest.fn(() => ({ addListener, removeListener }));
    window.matchMedia = matchMedia;

    const queries = {
      md: '(min-width: 801px) and (max-width: 900px)',
      lg: '(min-width: 901px)'
    };
    const component = mount(
      <Breakpoint queries={queries}>
        {({ md, lg }) => {
          if (lg) return <h1>Large</h1>;
          if (md) return <h1>Medium</h1>;
          return <h1>Small</h1>;
        }}
      </Breakpoint>
    );
    component.unmount();
    expect(removeListener).toHaveBeenCalledTimes(2);
  });

  it('renders with $expected query after evaluating queries with medium: $medium and large: $large', () => {
    /* eslint-disable no-undef, no-unused-labels, no-multi-spaces */
    const addListener = jest.fn();
    const matchMedia = jest
      .fn()
      .mockReturnValueOnce({ addListener, matches: medium })
      .mockReturnValueOnce({ addListener, matches: large });

    window.matchMedia = matchMedia;
    const component = mount(
      <Breakpoint queries={queries}>
        {({ md, lg }) => {
          if (lg) return <h1>Large</h1>;
          if (md) return <h1>Medium</h1>;
          return <h1>Small</h1>;
        }}
      </Breakpoint>
    );
    expect(component.html()).toContain(expected);

    where: {
      medium | large || expected;
      true | false || 'Medium';
      false | true || 'Large';
    }
  });
});
