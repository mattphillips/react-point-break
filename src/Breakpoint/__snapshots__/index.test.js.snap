// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breakpoint /> renders with given queries set to false 1`] = `
<Breakpoint
  queries={
    Object {
      "lg": "(min-width: 901px)",
      "md": "(min-width: 801px) and (max-width: 900px)",
    }
  }
>
  <h1>
    Small
  </h1>
</Breakpoint>
`;
